### PR TITLE
Optionally put API Gateways in GitLab VPC (#4001, #4168)

### PR DESCRIPTION
**Pull request by danielsotirhos**
Connected issue: #4001

### Author

- [x] Target branch is `develop`
- [x] Name of source branch matches `issues/<GitHub handle of author>/<issue#>-<slug>`
- [x] PR title references the connected issue
- [x] PR title matches<sup>1</sup> title of connected issue         <sub>or comment in PR explains why they're different</sub>
- [x] Title of at least one commit references connected issue
- [x] PR is connected to issue via Zenhub 
- [x] PR description links to connected issue
- [x] Added `partial` label to PR                                    <sub>or this PR completely resolves the connected issue</sub>

<sup>1</sup> when the issue title describes a problem, the PR title is `Fix: ` followed by the issue title   

### Author (reindex)

- [x] Added `r` tag to commit title                                 <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                                   <sub>or this PR does not require reindexing</sub>

### Author (chains)

- [x] This PR is blocked by previous PR in the chain               <sub>or this PR is not chained to another PR</sub>
- [x] Added `base` label to the blocking PR                        <sub>or this PR is not chained to another PR</sub>
- [x] Added `chained` label to this PR                             <sub>or this PR is not chained to another PR</sub>

### Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst          <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                                 <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                                   <sub>or this PR does not require upgrading</sub>

### Author (operator tasks)

- [x] Added checklist items for additional operator tasks           <sub>or this PR does not require additional tasks</sub>

### Author (hotfixes)

- [x] Added `F` tag to main commit title                            <sub>or this PR does not include permanent fix for a temporary hotfix</sub>
- [x] Reverted the temporary hotfix connected to the issue          <sub>or there is no temporary hotfix for the connected issue on the `prod` branch</sub>

### Author (requirements)

- [x] Ran `make requirements_update`                                <sub>or this PR does not touch requirements*.txt, common.mk, Makefile and Dockerfile</sub>
- [x] Added `R` tag to commit title                                 <sub>or this PR does not touch requirements*.txt</sub>
- [x] Added `reqs` label to PR                                      <sub>or this PR does not touch requirements*.txt</sub>

### Author (rebasing, integration test)

- [x] `make integration_test` passes in personal deployment         <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`, squashed old fixups

### Peer reviewer (after requesting changes)

Uncheck the *Author (requirements)* and *Author (rebasing, integration test)* 
checklists.

### Peer reviewer (after approval)

- [x] Ticket is in *Review requested* column
- [x] Requested review from primary reviewer
- [x] Assigned PR to primary reviewer

### Primary reviewer (after requesting changes)

Uncheck the *Author (requirements)* and *Author (rebasing, integration test)* 
checklists. Update the `N reviews` label. 

### Primary reviewer (after approval)

- [x] Actually approved the PR
- [x] Labeled connected issue as `demo` or `no demo`
- [x] Commented on connected issue about demo expectations          <sub>or labelled connected issue as `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] `N reviews` label is accurate
- [x] Moved ticket to *Approved* column
- [x] Assigned PR to current operator

### Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that demo expectations are clear                      <sub>or connected issue is labeled as `no demo`</sub>
- [x] Rebased and squashed branch, * removed `drop!` commit
- [x] Sanity-checked history
- [x] Pushed PR branch to GitHub
- [x] * Asked sys admin to deploy branch to GitLab for IAM permission updates.
- [x] Branch pushed to GitLab and added `sandbox` label             <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passed in sandbox                                       <sub>or PR is labeled `no sandbox`</sub>
- [x] Deleted unreferenced indices in `sandbox`                     <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices</sub> 
- [x] Started reindex in `sandbox`                                  <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox`                             <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [ ] Moved connected issue to Merged column
- [ ] Pushed merge commit to GitHub

### Operator (after pushing the merge commit)

- [ ] Shortened the PR chain                                        <sub>or this PR is not labeled `base`</sub>
- [ ] Pushed merge commit to GitLab                                 <sub>or merge commit can be pushed later, with another PR</sub>
- [ ] Deleted PR branch from GitHub and GitLab
- [ ] Build passes on GitLab

### Operator (reindex) 

- [ ] Deleted unreferenced indices in `dev`                         <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices</sub> 
- [ ] Started reindex in `dev`                                      <sub>or this PR does not require reindexing</sub>
- [ ] Checked for and triaged indexing failures                     <sub>or this PR does not require reindexing</sub>
- [ ] Emptied fail queues in target deployment                      <sub>or this PR does not require reindexing</sub>

### Operator

- [ ] Unassigned PR

## Shorthand for review comments

- `L` line is too long
- `W` line wrapping is wrong
- `Q` bad quotes
- `F` other formatting problem

### CODE REVIEW

File: terraform/gitlab/gitlab.tf.json.template.py [hunk 1]
```diff
@@ -662,6 +662,29 @@ def remove_inconsequential_statements(statements: List[JSON]) -> List[JSON]:
                 }
             }
         },
+        'aws_vpc_endpoint': {
```
Comment (hannes-ucsc) [inline terraform/gitlab/gitlab.tf.json.template.py:665]: I think we determined that we could share one VPC endpoint and this change is based on that determination. However, one endpoint costs 1 cent/h so $7.20/month. Knowing that private deployments will only be needed until we have compliance, and knowing that we'd be adding even more infrastructure to get custom domain names working, we should move these resources to the Azul deployment, outside of the GitLab component.

You will need to lookup the VPC and the subnets via TF data sources but create the endpoint and a separate security group in each deployment.

The demo scenario is likely going to be switching a personal deployment to private and back.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

**Comment by danielsotirhos** [comment]
azul-service-daniel: https://a5493cgt14-vpce-0340c1650794724ef.execute-api.us-east-1.amazonaws.com/daniel

File: src/azul/service/lambda_iam_policy.py [hunk 2]
```diff
@@ -181,6 +181,22 @@
                 }
             ] if direct_access_role is not None else [
             ]
+        ),
+        *(
```
Comment (hannes-ucsc) [inline src/azul/service/lambda_iam_policy.py:185]: Dedup into azul.chalice

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
Make sure IT passes and that you can revert your personal deployment between private and public.

```diff
Index: scripts/prepare_lambda_deployment.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/scripts/prepare_lambda_deployment.py b/scripts/prepare_lambda_deployment.py
--- a/scripts/prepare_lambda_deployment.py	(revision c59b534f15da48354d05ef24c2f3500dc689178c)
+++ b/scripts/prepare_lambda_deployment.py	(date 1652940696695)
@@ -34,9 +34,9 @@
         'es_endpoint': {},
         'es_instance_count': {},
         'cloudwatch_log_group_provisioner': {},
-        'vpc_endpoint_id': {},
-        'vpc_subnet_ids': {},
-        'vpc_security_group_id': {},
+        config.var_vpc_endpoint_id: {},
+        config.var_vpc_subnet_ids: {},
+        config.var_vpc_security_group_id: {},
     }
 
     input_json['output']['stage_name'] = {
@@ -48,7 +48,7 @@
         # add when the `api_gateway_endpoint_vpce` config is used.
         rest_api = input_json['resource']['aws_api_gateway_rest_api']['rest_api']
         rest_api['endpoint_configuration']['vpc_endpoint_ids'] = [
-            '${var.vpc_endpoint_id}'
+            '${var.%s}' % config.var_vpc_endpoint_id
         ]
 
     for func in input_json['resource']['aws_lambda_function'].values():
Index: lambdas/service/.chalice/config.json.template.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/lambdas/service/.chalice/config.json.template.py b/lambdas/service/.chalice/config.json.template.py
--- a/lambdas/service/.chalice/config.json.template.py	(revision c59b534f15da48354d05ef24c2f3500dc689178c)
+++ b/lambdas/service/.chalice/config.json.template.py	(date 1652940831955)
@@ -1,6 +1,9 @@
 from azul import (
     config,
 )
+from azul.chalice import (
+    private_api_config,
+)
 from azul.modules import (
     load_app_module,
 )
@@ -23,19 +26,7 @@
     "minimum_compression_size": config.minimum_compression_size,
     "lambda_timeout": config.api_gateway_lambda_timeout,
     "lambda_memory_size": 2048,
-    **(
-        {
-            "api_gateway_endpoint_type": "PRIVATE",
-            "api_gateway_endpoint_vpce": [
-                "${var.vpc_endpoint_id}"
-            ],
-            "subnet_ids": "${var.vpc_subnet_ids}",
-            "security_group_ids": [
-                "${var.vpc_security_group_id}"
-            ],
-        } if config.private_api else {
-        }
-    ),
+    **private_api_config(),
     "stages": {
         config.deployment_stage: {
             "lambda_functions": {
Index: lambdas/indexer/.chalice/config.json.template.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/lambdas/indexer/.chalice/config.json.template.py b/lambdas/indexer/.chalice/config.json.template.py
--- a/lambdas/indexer/.chalice/config.json.template.py	(revision c59b534f15da48354d05ef24c2f3500dc689178c)
+++ b/lambdas/indexer/.chalice/config.json.template.py	(date 1652940831963)
@@ -1,6 +1,9 @@
 from azul import (
     config,
 )
+from azul.chalice import (
+    private_api_config,
+)
 from azul.modules import (
     load_app_module,
 )
@@ -23,19 +26,7 @@
     "minimum_compression_size": config.minimum_compression_size,
     "lambda_timeout": config.api_gateway_lambda_timeout,
     "lambda_memory_size": 128,
-    **(
-        {
-            "api_gateway_endpoint_type": "PRIVATE",
-            "api_gateway_endpoint_vpce": [
-                "${var.vpc_endpoint_id}"
-            ],
-            "subnet_ids": "${var.vpc_subnet_ids}",
-            "security_group_ids": [
-                "${var.vpc_security_group_id}"
-            ],
-        } if config.private_api else {
-        }
-    ),
+    **private_api_config(),
     "stages": {
         config.deployment_stage: {
             "lambda_functions": {
Index: src/azul/chalice.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/chalice.py b/src/azul/chalice.py
--- a/src/azul/chalice.py	(revision c59b534f15da48354d05ef24c2f3500dc689178c)
+++ b/src/azul/chalice.py	(date 1652940696691)
@@ -358,3 +358,21 @@
     # Some type annotations to help with auto-complete
     lambda_context: LambdaContext
     current_request: AzulRequest
+
+
+def private_api_config():
+    """
+    Returns the fragment of Chalice configuration JSON that configures the
+    Lambda function to be invoked by a private API Gateway, if enabled.
+    """
+    return {
+        "api_gateway_endpoint_type": "PRIVATE",
+        "api_gateway_endpoint_vpce": [
+            "${var.%s}" % config.var_vpc_endpoint_id
+        ],
+        "subnet_ids": "${var.%s}" % config.var_vpc_subnet_ids,
+        "security_group_ids": [
+            "${var.%s}" % config.var_vpc_security_group_id
+        ],
+    } if config.private_api else {
+    }
Index: terraform/modules.tf.json.template.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/terraform/modules.tf.json.template.py b/terraform/modules.tf.json.template.py
--- a/terraform/modules.tf.json.template.py	(revision c59b534f15da48354d05ef24c2f3500dc689178c)
+++ b/terraform/modules.tf.json.template.py	(date 1652940696683)
@@ -27,13 +27,13 @@
                 if config.share_es_domain else
                 "${aws_elasticsearch_domain.index.cluster_config[0].instance_count}",
             "cloudwatch_log_group_provisioner": f"{lambda_name}_log_group_provisioner",
-            "vpc_endpoint_id": f"${{aws_vpc_endpoint.{lambda_name}_api_gateway.id}}",
-            "vpc_subnet_ids":
+            config.var_vpc_endpoint_id: f"${{aws_vpc_endpoint.{lambda_name}_api_gateway.id}}",
+            config.var_vpc_subnet_ids:
                 [
                     f"${{data.aws_subnet.gitlab_{vpc.subnet_name(public=True)}_{zone}.id}}"
                     for zone in range(vpc.num_zones)
                 ],
-            "vpc_security_group_id": f"${{aws_security_group.{lambda_name}.id}}"
+            config.var_vpc_security_group_id: f"${{aws_security_group.{lambda_name}.id}}"
         } for lambda_name in config.lambda_names()
     }
 })
Index: src/azul/__init__.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/__init__.py b/src/azul/__init__.py
--- a/src/azul/__init__.py	(revision c59b534f15da48354d05ef24c2f3500dc689178c)
+++ b/src/azul/__init__.py	(date 1652940696699)
@@ -1018,6 +1018,12 @@
 
     permissions_boundary_name = 'azul-boundary'
 
+    var_vpc_endpoint_id = 'vpc_endpoint_id'
+
+    var_vpc_subnet_ids = 'vpc_subnet_ids'
+
+    var_vpc_security_group_id = 'vpc_security_group_id'
+
     @property
     def github_project(self) -> str:
         return self.environ['azul_github_project']
```

Prefer `%` when interpolating into strings that already contain curly braces.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
```diff
Index: README.md
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/README.md b/README.md
--- a/README.md	(revision 74f3f1ebf9f07d4f6429ecb406c09567a33fe66d)
+++ b/README.md	(date 1654107705782)
@@ -663,43 +663,47 @@
 
 ## 3.6 Private API
 
-Follow these steps to put a deployment's API Gateway behind the GitLab VPC. Once
-completed, a VPN connection will be required to access the deployment. See
+Follow these steps to put a deployment's API Gateway in the GitLab VPC so that a 
+VPN connection is required to access the deployment. See
 [9.1 VPN access to GitLab](#91-vpn-access-to-gitlab) for details.
 
-1. Destroy the current deployment (`make -C terraform destroy;`).
+1. Destroy the current deployment (`make -C terraform destroy`).
 2. Increment `AZUL_DEPLOYMENT_INCARNATION`.
 3. Set `AZUL_PRIVATE_API` to `1`.
-4. Deploy (`make deploy`).
+4. Redeploy (`make deploy`).
 
-Troubleshooting:
+### Troubleshooting
 
 Transient errors might be encountered during the deploy such as `SQS Error Code:
 AWS.SimpleQueueService.NonExistentQueue. SQS Error Message: The specified queue
-does not exist for this wsdl version.` In such cases rerunning the deploy step
+does not exist for this wsdl version` In such cases rerunning `make deploy`
 should resolve the issue.
 
 [aws_cloudwatch_log_group]: https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/cloudwatch_log_group
 
 If the error `ResourceAlreadyExistsException: The specified log group already
 exists` is encountered, import the [aws_cloudwatch_log_group] resource into
-terraform and retry the deploy.
+terraform and retry the deploy. REVIEW: how?
 
 If the error `azul.RequirementError: The service account (SA) '...' is not
 authorized to access ... or that resource does not exist. Make sure that it 
 exists, that the SA is registered with SAM and has been granted read access to 
-the resource.` is encountered, ask an administrator of the TDR `azul-dev` group
-to add your service account specified in the error messaged to the group. See
+the resource` is encountered, ask an administrator of the Terra group `azul-dev` 
+to add the service account as specified in the error messaged to that group. See
 [2.3.4 Google Cloud, TDR, and SAM](#234-google-cloud-tdr-and-sam) for details.
 
 [KMSAccessDeniedException]: https://aws.amazon.com/premiumsupport/knowledge-center/lambda-kmsaccessdeniedexception-errors/
 
-After the deploy, if the deployment is unresponsive and CloudWatch reports logs
-in the `/aws/apigateway` Log group but not in `/aws/lambda/`, you can confirm
-the issue is a [KMSAccessDeniedException] error via the AWS Console by going to
-the Lambda function details page, clicking on the `Test` tab, and clicking on
-the `Test` buttton. With the deployment selected, the `reset_lambda_role.py`
-script can be used to reset all Lamda functions to resolve this issue.
+After a successful invocation of `make deploy`, if the deployment is unresponsive and CloudWatch shows logs entries
+in the `/aws/apigateway/…` log group but not in `/aws/lambda/…`, first confirm
+whether the issue is the known [KMSAccessDeniedException] error. In the AWS Console, go to
+the Lambda function details page, click on the `Test` tab, and click on
+the `Test` buttton. 
+
+REVIEW: Readers might wrongly interpret any error as a symptom of the issue. Clarify that. 
+
+If confirmed, and with the target deployment selected, run the `reset_lambda_role.py`
+script to reset all Lambda functions and resolve this issue.
 
 ## 3.7 Reindexing
 
```

**Comment by danielsotirhos** [comment]
Deployment with APIs behind the GitLab VPC: https://service.daniel2.dev.singlecell.gi.ucsc.edu/

File: scripts/reset_lambda_role.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+)
+from azul.logging import (
+    configure_script_logging,
+)
+
+log = logging.getLogger(__name__)
+
+
+def main():
+    client = boto3.client('lambda')
```
Comment (noah-aviel-dove) [inline scripts/reset_lambda_role.py:20]:
```suggestion
    client = aws.lambda_
```

File: scripts/reset_lambda_role.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+    paginator = client.get_paginator('list_functions')
+
+    response_iterator = paginator.paginate()
+    lambda_names = set(config.lambda_names())
+
+    for page in response_iterator:
```
Comment (noah-aviel-dove) [inline scripts/reset_lambda_role.py:27]: Consider using `azul.lambdas.Lambdas.list_lambdas` and moving some/all of this functionality into that module.

File: README.md [hunk 5]
```diff
@@ -629,7 +629,7 @@ make create
```
Comment (noah-aviel-dove) [inline README.md:629]: I don't think this should be a separate commit. The changes to the documentation should be in the same commit(s) as the corresponding code changes. Same goes for UPGRADING.rst.

File: terraform/api_gateway.tf.json.template.py [hunk 6]
```diff
@@ -241,7 +290,9 @@ def for_domain(cls, domain):
                             )
                         ]
                     }),
-                    **aws.permissions_boundary_tf
+                    # Temporally disabled until changes in this branch have been
+                    # deployed to GitLab to extend the boundary permissions.
+                    # **aws.permissions_boundary_tf
```
Comment (noah-aviel-dove) [inline terraform/api_gateway.tf.json.template.py:295]: We don't use comments for disabling code.

```suggestion
                    # Temporally disabled until changes in this branch have been
                    # deployed to GitLab to extend the boundary permissions.
                    **(aws.permissions_boundary_tf if False else {})
```

**Review by noah-aviel-dove — CHANGES_REQUESTED** [review]
It would be nice to see green builds for the individual commits given the long history.

File: terraform/Makefile [hunk 7]
```diff
@@ -30,10 +30,16 @@ plan: validate
...
+	$(if $(findstring 1,$(AZUL_PRIVATE_API)), \
+	terraform apply -target aws_vpc_endpoint.indexer -target aws_vpc_endpoint.service \
+	)
 	terraform apply
 
 .PHONY: auto_apply
 auto_apply: validate rename_resources
+	$(if $(findstring 1,$(AZUL_PRIVATE_API)),\
+		terraform apply -auto-approve -target aws_vpc_endpoint.indexer -target aws_vpc_endpoint.service \
+	)
```
[the remark this answers is not in the document]
Comment (danielsotirhos) [reply terraform/Makefile:42]: These resources need to be created first so that the `for_each` in the `aws_network_interface` data source (api_gateway.tf.json.template.py) will be able to iterate the network interface ids of the `aws_vpc_endpoint`s.

File: lambdas/indexer/.chalice/config.json.template.py [hunk 8]
```diff
@@ -25,7 +29,11 @@
     "lambda_memory_size": 128,
     "stages": {
         config.deployment_stage: {
+            **private_api_stage_config(),
             "lambda_functions": {
+                "api_handler": {
+                    **private_api_lambda_config()
+                },
```
Comment (noah-aviel-dove) [inline lambdas/indexer/.chalice/config.json.template.py:36]:
```suggestion
                "api_handler": private_api_lambda_config(),
```

File: lambdas/service/.chalice/config.json.template.py [hunk 9]
```diff
@@ -25,7 +29,11 @@
     "lambda_memory_size": 2048,
     "stages": {
         config.deployment_stage: {
+            **private_api_stage_config(),
             "lambda_functions": {
+                "api_handler": {
+                    **private_api_lambda_config()
+                },
```
Comment (noah-aviel-dove) [inline lambdas/service/.chalice/config.json.template.py:36]:
```suggestion
                "api_handler": private_api_lambda_config(),
```

File: terraform/api_gateway.tf.json.template.py (hunk 6 above)
Comment (noah-aviel-dove) [reply terraform/api_gateway.tf.json.template.py]: Changes in the `fixup` do not match the suggestion.

**Review by noah-aviel-dove — APPROVED** [review]

File: README.md [hunk 10]
```diff
@@ -661,7 +661,44 @@ http://indexer.${AZUL_DEPLOYMENT_STAGE}.dev.singlecell.gi.ucsc.edu/
...
+does not exist for this wsdl version` In such cases rerunning `make deploy`
+should resolve the issue.
+
+[aws_cloudwatch_log_group]: https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/cloudwatch_log_group
+
+If the error `ResourceAlreadyExistsException: The specified log group already
+exists` is encountered, follow the steps below to import the
+[aws_cloudwatch_log_group] resources into terraform and retry the deploy.
+
+1. `cd terraform`
```
Comment (hannes-ucsc) [inline README.md:688]: same here

File: src/azul/chalice.py [hunk 11]
```diff
@@ -351,3 +351,51 @@ def load_resource(self, *path: str) -> str:
...
+    }
+
+
+def private_api_policy(for_tf: bool = False):
+    """
+    Returns the fragment of IAM policy JSON needed for placing a Lambda function
+    into a VPC so that it can be invoked by a private API Gateway, if enabled.
+    """
+    return [
+        {} if for_tf else {'Effect': 'Allow'} | {
```
Comment (hannes-ucsc) [inline src/azul/chalice.py:392]: Is `|` really necessary? It looks like it's joining two literals which would make it redundant.

File: terraform/Makefile (hunk 7 above)
[the remark this answers is not in the document]
Comment (hannes-ucsc) [reply terraform/Makefile:42]: [This](https://github.com/DataBiosphere/azul/pull/4144/commits/1a711d111053ca4a3073e1f2ced2647201ed338b#r894768964) should be explained in a comment here and in the file section referenced by your answer to the reviewers question.

File: terraform/Makefile [hunk 12]
```diff
@@ -30,10 +30,16 @@ plan: validate
 
 .PHONY: apply
 apply: validate rename_resources
+	$(if $(findstring 1,$(AZUL_PRIVATE_API)), \
```
Comment (hannes-ucsc) [inline terraform/Makefile:33]: The use of findstring is potentially dangerous. Is there no equality operator in GNU make? Look for `ifeq`

File: terraform/Makefile [hunk 13]
```diff
@@ -30,10 +30,16 @@ plan: validate
...
+	terraform apply -target aws_vpc_endpoint.indexer -target aws_vpc_endpoint.service \
```
Comment (hannes-ucsc) [inline terraform/Makefile:34]: Doesn't match indent of the next recipe.

File: terraform/api_gateway.tf.json.template.py [hunk 14]
```diff
@@ -375,7 +375,9 @@ def for_domain(cls, domain):
                                 "unhealthy_threshold": 2,
                                 "timeout": 5,
                                 "interval": 30,
-                                "matcher": "200"
+                                # FIXME: Health check for private API requires 403
```
Comment (hannes-ucsc) [inline terraform/api_gateway.tf.json.template.py:380]: Didn't we close that issue with`wontfix`?

File: README.md [hunk 15]
```diff
@@ -706,11 +706,12 @@ and CloudWatch shows logs entries in the `/aws/apigateway/…` log group but not
 [KMSAccessDeniedException] error. In the AWS Console, go to the Lambda function
 details page, click on the `Test` tab, and click on the `Test` buttton. 
 
-Note that is normal for some Lambda to fail the test due to the parameters of
-the test event. Examine the error message to determine if the failure is due to
-a KMSAccessDeniedException which would be explicitly specified.
+Note that is normal for some Lambda functions to fail the test due to the
```
Comment (hannes-ucsc) [inline README.md:709]: "that it is normal"

File: README.md [hunk 16]
```diff
@@ -706,11 +706,12 @@ and CloudWatch shows logs entries in the `/aws/apigateway/…` log group but not
...
+parameters of the test event. Examine the error message to determine if the
+failure is due to a `KMSAccessDeniedException` which would be explicitly
+specified.
 
-If a KMSAccessDeniedException is confirmed, with the target deployment
+If a `KMSAccessDeniedException` is confirmed, with the target deployment
```
Comment (hannes-ucsc) [inline README.md:714]: "If `KMSAccessDeniedException` is confirmed to be the cause"

comma between selected and run. But still awkward. Please improve.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
No fixups please next time around.

File: terraform/api_gateway.tf.json.template.py [hunk 17]
```diff
@@ -79,14 +80,55 @@ def for_domain(cls, domain):
...
+                                "values": [f"azul-gitlab-{vpc.subnet_name(public)}-{zone}"]
+                            }
+                        }
+                        for public in (False, True)
+                        for zone in range(vpc.num_zones)
+                    }
+                },
+                {
+                    # In order for this to work, the VPC endpoints must have been
+                    # created before the network interface ids can be iterated here.
```
Comment (hannes-ucsc) [inline terraform/api_gateway.tf.json.template.py:121]: This needs to reference the Makefile otherwise it would leave the reader extremely confused.

```suggestion
                    # In order for this to work, the VPC endpoints must have been
                    # created before the network interface IDs can be iterated here.
```

An id is something else.

https://www.merriam-webster.com/dictionary/id

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

**Review by hannes-ucsc — APPROVED** [review]
There should have been a fixup commit. The request to omit fixups was for the penultimate review cycle only.

**Review by hannes-ucsc — APPROVED** [review]
